### ISSUE TITLE
rw gce pd mounted as ro

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#27058]
I have a pvc like:
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.alpha.kubernetes.io/storage-class: anything
  creationTimestamp: 2016-06-08T07:06:29Z
  labels:
    app: zk
  name: datadir-zoo-0
  namespace: e2e-tests-petset-7zm2s
  resourceVersion: "182261"
  selfLink: /api/v1/namespaces/e2e-tests-petset-7zm2s/persistentvolumeclaims/datadir-zoo-0
  uid: 85da26bf-2d47-11e6-8c15-42010af00002
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  volumeName: pvc-85da26bf-2d47-11e6-8c15-42010af00002
status:
  phase: Bound
```

It created a r/w pd (verified through the pantheon console) but the actual disk is mounted as ro.
```
root@e2e-test-beeps-minion-group-nlnf:/var/lib/kubelet/pods/c86d8439-2d47-11e6-8c15-42010af00002/volumes# mount | grep pvc-85da26bf-2d47-11e6-8c15-42010af00002
/dev/sdf on /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/e2e-test-beeps-dynamic-pvc-85da26bf-2d47-11e6-8c15-42010af00002 type ext4 (ro,relatime,data=ordered)
/dev/sdf on /var/lib/kubelet/pods/c86d8439-2d47-11e6-8c15-42010af00002/volumes/kubernetes.io~gce-pd/pvc-85da26bf-2d47-11e6-8c15-42010af00002 type ext4 (ro,relatime,data=ordered)

root@e2e-test-beeps-minion-group-nlnf:/var/lib/kubelet/pods/c86d8439-2d47-11e6-8c15-42010af00002/volumes# ls kubernetes.io~gce-pd/pvc-85da26bf-2d47-11e6-8c15-42010af00002/
ls: reading directory kubernetes.io~gce-pd/pvc-85da26bf-2d47-11e6-8c15-42010af00002/: Input/output error

```

Similar disks created previously are mounted as r/w on the same vm.
```
root@e2e-test-beeps-minion-group-nlnf:/var/lib/kubelet/pods/527ec9be-2d3f-11e6-8c15-42010af00002/volumes# mount -v | grep pvc-a418664d-2d38-11e6-8c15-42010af00002
/dev/sde on /var/lib/kubelet/plugins/kubernetes.io/gce-pd/mounts/e2e-test-beeps-dynamic-pvc-a418664d-2d38-11e6-8c15-42010af00002 type ext4 (rw,relatime,data=ordered)
/dev/sde on /var/lib/kubelet/pods/527ec9be-2d3f-11e6-8c15-42010af00002/volumes/kubernetes.io~gce-pd/pvc-a418664d-2d38-11e6-8c15-42010af00002 type ext4 (rw,relatime,data=ordered)

```

Oddly i couldn't find the partition identifier when i tried to remount it manually
```
root@e2e-test-beeps-minion-group-nlnf:/var/lib/kubelet/pods/c86d8439-2d47-11e6-8c15-42010af00002/volumes/kubernetes.io~gce-pd# ls /dev/sdf
ls: cannot access /dev/sdf: No such file or directory

root@e2e-test-beeps-minion-group-nlnf:/var/lib/kubelet/pods/527ec9be-2d3f-11e6-8c15-42010af00002/volumes# ls /dev/sde
/dev/sde
```

My pod will keep crashlooping till the volume is r/w, how to recover (without deleting pod)?
@kubernetes/sig-storage

**Comment by saad-ali** [comment]
> It created a r/w pd (verified through the pantheon console) but the actual disk is mounted as ro.

Can you post the PV and Pod objects you are using with the PVC.

**Comment by bprashanth** [comment]
Ignore the weird json blob it's an init container:
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubernetes.io/createdby: gce-pd-dynamic-provisioner
    pv.kubernetes.io/bound-by-controller: "yes"
    pv.kubernetes.io/provisioned-by: kubernetes.io/gce-pd
  creationTimestamp: 2016-06-08T07:06:32Z
  labels:
    failure-domain.beta.kubernetes.io/region: us-central1
    failure-domain.beta.kubernetes.io/zone: us-central1-b
  name: pvc-85da26bf-2d47-11e6-8c15-42010af00002
  resourceVersion: "182259"
  selfLink: /api/v1/persistentvolumes/pvc-85da26bf-2d47-11e6-8c15-42010af00002
  uid: 880af5b5-2d47-11e6-8c15-42010af00002
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: datadir-zoo-0
    namespace: e2e-tests-petset-7zm2s
    resourceVersion: "182248"
    uid: 85da26bf-2d47-11e6-8c15-42010af00002
  gcePersistentDisk:
    pdName: e2e-test-beeps-dynamic-pvc-85da26bf-2d47-11e6-8c15-42010af00002
  persistentVolumeReclaimPolicy: Delete
status:
  phase: Bound
---
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/created-by: |
      {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"PetSet","namespace":"e2e-tests-petset-7zm2s","name":"zoo","uid":"85d93a53-2d47-11e6-8c15-42010af00002","apiVersion":"apps","resourceVersion":"182372"}}
    pod.alpha.kubernetes.io/init-container-statuses: '[{"name":"bootstrap","state":{"waiting":{"reason":"CrashLoopBackOff","message":"Back-off
      5m0s restarting failed container=bootstrap pod=zoo-0_e2e-tests-petset-7zm2s(c86d8439-2d47-11e6-8c15-42010af00002)"}},"lastState":{"terminated":{"exitCode":1,"reason":"Error","startedAt":"2016-06-08T20:12:05Z","finishedAt":"2016-06-08T20:12:05Z","containerID":"docker://87e7e99a69263c2887d313569eb0196cac6f2675dff02b1a2cab075ff4389e1a"}},"ready":false,"restartCount":158,"image":"java:openjdk-8-jre","imageID":"docker://sha256:f707bac8cb4816c94c7ae3259d3803e161edcbb52a450e02941cbf539e7132e2","containerID":"docker://87e7e99a69263c2887d313569eb0196cac6f2675dff02b1a2cab075ff4389e1a"},{"name":"install","state":{"terminated":{"exitCode":0,"reason":"Completed","startedAt":"2016-06-08T07:08:23Z","finishedAt":"2016-06-08T07:08:23Z","containerID":"docker://c7034d47dd517d2e15960a4b5a0a5823f71ce97ef24c5604e25bc1f7dbcc9d42"}},"lastState":{},"ready":true,"restartCount":0,"image":"bprashanth/zookeeper-install-3.5.0-alpha:e2e","imageID":"docker://sha256:f415c9930a0f280c70c8c5728deb3692c3982f5db2499d627c6e44ce96fb3679","containerID":"docker://c7034d47dd517d2e15960a4b5a0a5823f71ce97ef24c5604e25bc1f7dbcc9d42"}]'
    pod.alpha.kubernetes.io/init-containers: '[{"name":"install","image":"bprashanth/zookeeper-install-3.5.0-alpha:e2e","args":["--install-into=/opt","--work-dir=/work-dir"],"resources":{},"volumeMounts":[{"name":"opt","mountPath":"/opt/"},{"name":"workdir","mountPath":"/work-dir"},{"name":"default-token-71uep","readOnly":true,"mountPath":"/var/run/secrets/kubernetes.io/serviceaccount"}],"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"Always"},{"name":"bootstrap","image":"java:openjdk-8-jre","command":["/work-dir/peer-finder"],"args":["-on-start=\"/work-dir/on-start.sh\"","-service=zk"],"env":[{"name":"POD_NAMESPACE","valueFrom":{"fieldRef":{"apiVersion":"v1","fieldPath":"metadata.namespace"}}}],"resources":{},"volumeMounts":[{"name":"opt","mountPath":"/opt/"},{"name":"workdir","mountPath":"/work-dir"},{"name":"datadir","mountPath":"/tmp/zookeeper"},{"name":"default-token-71uep","readOnly":true,"mountPath":"/var/run/secrets/kubernetes.io/serviceaccount"}],"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"IfNotPresent"}]'
    pod.alpha.kubernetes.io/initialized: "true"
    pod.beta.kubernetes.io/hostname: zoo-0
    pod.beta.kubernetes.io/subdomain: zk
  creationTimestamp: 2016-06-08T07:08:20Z
  generateName: zoo-
  labels:
    app: zk
  name: zoo-0
  namespace: e2e-tests-petset-7zm2s
  resourceVersion: "202683"
  selfLink: /api/v1/namespaces/e2e-tests-petset-7zm2s/pods/zoo-0
  uid: c86d8439-2d47-11e6-8c15-42010af00002
spec:
  containers:
  - args:
    - start-foreground
    command:
    - /opt/zookeeper/bin/zkServer.sh
    image: java:openjdk-8-jre
    imagePullPolicy: IfNotPresent
    name: zk
    ports:
    - containerPort: 2888
      name: peer
      protocol: TCP
    - containerPort: 3888
      name: leader-election
      protocol: TCP
    readinessProbe:
      exec:
        command:
        - sh
        - -c
        - /opt/zookeeper/bin/zkCli.sh ls /
      failureThreshold: 3
      initialDelaySeconds: 15
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 5
    resources: {}
    terminationMessagePath: /dev/termination-log
    volumeMounts:
    - mountPath: /tmp/zookeeper
      name: datadir
    - mountPath: /opt/
      name: opt
    - mountPath: /work-dir
      name: workdir
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: default-token-71uep
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: e2e-test-beeps-minion-group-nlnf
  restartPolicy: Always
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 0
  volumes:
  - name: datadir
    persistentVolumeClaim:
      claimName: datadir-zoo-0
  - emptyDir: {}
    name: opt
  - emptyDir: {}
    name: workdir
  - name: default-token-71uep
    secret:
      secretName: default-token-71uep
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2016-06-08T07:08:21Z
    message: 'containers with incomplete status: [bootstrap]'
    reason: ContainersNotInitialized
    status: "False"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2016-06-08T07:08:21Z
    message: 'containers with unready status: [zk]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2016-06-08T07:08:21Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: java:openjdk-8-jre
    imageID: ""
    lastState: {}
    name: zk
    ready: false
    restartCount: 0
    state:
      waiting:
        reason: PodInitializing
  hostIP: 10.240.0.3
  phase: Pending
  podIP: 10.180.1.8
  startTime: 2016-06-08T07:08:21Z
```

I haven't debugged enough to say this is not "kubelet doing somethign different for init containers", but it works 90% of the time.

**Comment by saad-ali** [comment]
The PV and Pod objects look fine. Spoke with @bprashanth offline. Looks like this happens randomly not on every execution. He'll add extra logging to gather more details.

**Comment by bprashanth** [comment]
I'll reopen when i have the requested logs. I've also modified my test in a way that might be hiding this failure mode, it failed for other reasons in the last overnight run.

**Comment by bprashanth** [comment]
This happens when a pod gets re-scheduled to a node from which the attacher is detaching a pv mounted by the pod. We end up with:

1. detach start
2. pod lands on node
3. kubelet checks to see that pv is attached
4. detach completes
5. attach start
6. attach complete

In between 4 and 6 the pod doesn't have a usable volume:
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-petset/165/artifacts
 
Reopening. Saad is on the case. Would be nice to try and fix this for 1.3 since it causes flake in petset e2es daily. Let me know if I can help.

**Comment by saad-ali** [comment]
I noticed this while testing https://github.com/kubernetes/kubernetes/pull/26801 as well. I have a plan for how to fix it, and agreed this needs to be fixed for 1.3.

**Comment by saad-ali** [comment]
Closing this issue in favor of https://github.com/kubernetes/kubernetes/issues/27483
